### PR TITLE
Abstract domains for bitwise operations

### PR DESCRIPTION
**Pull request by robdockins**
Follow on to PR #387

### CODE REVIEW

**Comment by robdockins** [comment]
@brianhuffman, If CI comes back green, I think this is ready to review and merge.

File: what4/src/What4/Expr/Builder.hs [hunk 1]
```diff
@@ -334,6 +345,53 @@ data NonceApp t (e :: BaseType -> Type) (tp :: BaseType) where
         -> !(Ctx.Assignment e args)
         -> NonceApp t e ret
 
+
+-------------------------------------------------------------------------------
+-- BVOrSet
+
+data BVOrNote w = BVOrNote !IncrHash !(BVD.BVDomain w)
```
Comment (brianhuffman) [inline what4/src/What4/Expr/Builder.hs]: It would be nice to use a special bit-oriented abstract domain here instead of the normal contiguous-range-of-bitvectors one. But it probably makes sense to do that as a separate PR.

Comment (robdockins) [reply what4/src/What4/Expr/Builder.hs]: Agreed, on both counts.

File: what4/src/What4/Expr/WeightedSum.hs [hunk 2]
```diff
@@ -92,16 +92,13 @@ data SRAbsValue :: SR.SemiRing -> Type where
...
-  SRAbsBVOr    :: (1 <= w) => NatRepr w      -> SRAbsValue (SR.SemiRingBV SR.BVBits w)
 
 instance Semigroup (SRAbsValue sr) where
   SRAbsNatAdd  x <> SRAbsNatAdd  y = SRAbsNatAdd  (AD.natRangeAdd x y)
   SRAbsIntAdd  x <> SRAbsIntAdd  y = SRAbsIntAdd  (AD.addRange x y)
   SRAbsRealAdd x <> SRAbsRealAdd y = SRAbsRealAdd (AD.ravAdd x y)
   SRAbsBVAdd   x <> SRAbsBVAdd   y = SRAbsBVAdd   (BVD.add x y)
   SRAbsBVXor   w <> SRAbsBVXor   _ = SRAbsBVXor w
-  SRAbsBVOr    w <> SRAbsBVOr    _ = SRAbsBVOr w
-  _              <> _              = error "internal error: WeightedSum abstract domain mismatch"
```
[the remark this answers is not in the document]
Comment (robdockins) [reply what4/src/What4/Expr/WeightedSum.hs]: Indeed, GHC correctly deduces that this is now a complete pattern match.
